### PR TITLE
Some Grain calls throw SiloUnavailableException when they should be successful 

### PR DESCRIPTION
**Pull request by benjaminpetit**
Proposed fix for #2282

### CODE REVIEW

File: src/Orleans/Messaging/GatewayManager.cs [hunk 1]
```diff
@@ -134,6 +134,10 @@ public Uri GetLiveGateway()
             {
                 lock (lockable)
                 {
+                    if (config.PreferedGatewayIndex >= 0 && config.PreferedGatewayIndex < count)
+                    {
+                        return live[config.PreferedGatewayIndex];
```
Comment (jdom) [inline src/Orleans/Messaging/GatewayManager.cs]: Already talked in person with @benjaminpetit, but we can't do this... He'll have the tests call several times until there's a hide confidence that this path will be hit.

Comment (benjaminpetit) [inline src/Orleans/Messaging/GatewayManager.cs]: Fixed!

**Comment by gabikliot** [comment]
I am not sure about this "fix".

First, there are 2 different issues here: 
1) msg is queued on the sender client (lets talk about client only for now) but was not sent yet vs. was already sent.
2) the gateway through which it was sent failed, but not the silo where the msg was destined to.

Your test checks 2, while the fix (TentativeCount) addresses  1 (if TentativeCount>1, do not break, otherwise [queued  but not sent] dobreak).

Issue 1: I think the issue of queued but not sent is a non issue - there is a short period between when we pick the GW and when we actually sent the msg (this queued period), so I would not address that at all. Just ignore that short period. If the msg is created but GW is not picked yet, we will not route it via dead GW anyway.

Issue 2: GW went down but you still do not want to break the msg. Well, if that is what you want, just don't do OnTargetSiloFail from BreakOutstandingMessagesToDeadSilo in OutsideRuntimeClient..
BUT, the problem is: you don't know on the client who is the real destination. It may be this GW or maybe another silo. So if you pick to not break, you might be waiting the full 30sec since the silo actually did fail.

So anyway, I don't think the prev. behavior is a bug. If you prefer just not to break msgs when GW fails, since the msg may have been re-routed to a different silo, I am OK with that. But then you don't need the TentativeCount. Simply do not break the msg.

**Comment by benjaminpetit** [comment]
Well I added the TentativeCount to be able to break with an exception only when the message was not sent but I think I can find a more elegant way to do it. And after thoughts, it seems that it would be better/easier to fix #2283 and #2285 on the same commit: remove the `BreakOutstandingMessagesToDeadSilo` method and reroute all message not yet sent to another GW when possible.

On the contrary, to me it seems like a bug:

- In client -> silo case: there is a good chance that the grain you were targeting was not hosted on the silo you used as a gateway. Your call has a good chance to be successful, even if the gateway was not gracefully stopped.
- In silo -> silo case: I also think we should not fail and try to reroute the call, even if we were targeting the correct silo, because ownership might have been transferred to another silo.

To me, we should only throw `SiloUnavailableException` when the request was sent to the correct silo and that this silo was not able to process or forward it. Or when the gateway you used was not able to forward the request before stopping. But I am not sure this is doable...

**Comment by gabikliot** [comment]
I agree we should try to resend any msg that we can, but only if we are 100% sure it didn't start processing yet.  For msgs we are not sure, we can't resend. And then, for those msgs we have 2 choices: pessimisticslly break early or optimistically hope to get an answer and if not time out.

The first case ( not resend when we can) is a bug. The second case ( pessimistic or optimistic) is a policy decision. I agree with you that in most cases optimistic makes more sense.

**Comment by benjaminpetit** [comment]
Closing this one because it is partially fixed in #2298. Let's discuss in #2282 if we should fix it for inter silos communication

**Pull request closed without merging**